src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -1387,26 +1401,32 @@ public SocketError SendToAsync(IList<ArraySegment<byte>> buffers, SocketFlags fl
 
         public unsafe void HandleEvents(Interop.Sys.SocketEvents events)
         {
-            lock (_queueLock)
+            if ((events & Interop.Sys.SocketEvents.Error) != 0)
             {
-                if ((events & Interop.Sys.SocketEvents.Error) != 0)
-                {
-                    // Set the Read and Write flags as well; the processing for these events
-                    // will pick up the error.
-                    events |= Interop.Sys.SocketEvents.Read | Interop.Sys.SocketEvents.Write;
-                }
+                // Set the Read and Write flags as well; the processing for these events
+                // will pick up the error.
+                events |= Interop.Sys.SocketEvents.Read | Interop.Sys.SocketEvents.Write;
+            }
 
-                if ((events & Interop.Sys.SocketEvents.Read) != 0)
+            if ((events & Interop.Sys.SocketEvents.Read) != 0)
+            {
+                lock (_sendAcceptConnectLock)
                 {
                     if (_acceptOrConnectQueue.AllOfType<AcceptOperation>())
                     {
                         _acceptOrConnectQueue.Complete(this);
                     }
+                }
 
+                lock (_receiveLock)
```

Comment:
Agreed, but I'm not sure how to avoid this without a much larger change (#6912).